File: backend/routes/address.js
const express = require('express');
const router = express.Router();
const { protect } = require('../middleware/auth');

// Search addresses using Google Maps Places Autocomplete API
router.get('/search', async (req, res) => {
  try {
    const { query } = req.query;
    
    if (!query) {
      return res.status(400).json({ message: 'Query parameter is required' });
    }

    const apiKey = process.env.GOOGLE_MAPS_API_KEY;
    if (!apiKey) {
      return res.status(500).json({ message: 'Google Maps API key not configured' });
    }

    // Use Places Autocomplete API
    const response = await fetch(
      `https://maps.googleapis.com/maps/api/place/autocomplete/json?input=${encodeURIComponent(query)}&key=${apiKey}&components=country:in`
    );
    
    const data = await response.json();
    
    if (data.status === 'OK') {
      res.json({ predictions: data.predictions });
    } else {
      res.json({ predictions: [] });
    }
  } catch (error) {
    console.error('Address search error:', error);
    res.status(500).json({ message: 'Error searching addresses' });
  }
});

// Geocode address to get coordinates
router.post('/geocode', async (req, res) => {
  try {
    const { address } = req.body;
    
    if (!address) {
      return res.status(400).json({ message: 'Address is required' });
    }

    const apiKey = process.env.GOOGLE_MAPS_API_KEY;
    if (!apiKey) {
      return res.status(500).json({ message: 'Google Maps API key not configured' });
    }

    // Geocode the address
    const response = await fetch(
      `https://maps.googleapis.com/maps/api/geocode/json?address=${encodeURIComponent(address)}&key=${apiKey}`
    );
    
    const data = await response.json();
    
    if (data.status === 'OK' && data.results.length > 0) {
      const result = data.results[0];
      const { lat, lng } = result.geometry.location;
      
      res.json({
        latitude: lat,
        longitude: lng,
        formattedAddress: result.formatted_address
      });
    } else {
      res.status(404).json({ message: 'Address not found' });
    }
  } catch (error) {
    console.error('Geocoding error:', error);
    res.status(500).json({ message: 'Error geocoding address' });
  }
});

// Validate address using Google Maps Geocoding API
router.post('/validate', protect, async (req, res) => {
  try {
    const { address } = req.body;
    
    if (!address) {
      return res.status(400).json({ message: 'Address is required' });
    }

    const apiKey = process.env.GOOGLE_MAPS_API_KEY;
    if (!apiKey) {
      return res.status(500).json({ message: 'Google Maps API key not configured' });
    }

    // Geocode the address
    const response = await fetch(
      `https://maps.googleapis.com/maps/api/geocode/json?address=${encodeURIComponent(address)}&key=${apiKey}`
    );
    
    const data = await response.json();
    
    if (data.status === 'OK' && data.results.length > 0) {
      const result = data.results[0];
      const { lat, lng } = result.geometry.location;
      
      // Extract address components
      const addressComponents = result.address_components;
      const streetNumber = addressComponents.find(component => 
        component.types.includes('street_number')
      )?.long_name || '';
      
      const route = addressComponents.find(component => 
        component.types.includes('route')
      )?.long_name || '';
      
      const locality = addressComponents.find(component => 
        component.types.includes('locality')
      )?.long_name || '';
      
      const postalCode = addressComponents.find(component => 
        component.types.includes('postal_code')
      )?.long_name || '';
      
      const country = addressComponents.find(component => 
        component.types.includes('country')
      )?.long_name || '';
      
      res.json({
        valid: true,
        coordinates: { latitude: lat, longitude: lng },
        formattedAddress: result.formatted_address,
        addressComponents: {
          street: `${streetNumber} ${route}`.trim(),
          city: locality,
          pincode: postalCode,
          country: country
        }
      });
    } else {
      res.json({
        valid: false,
        message: 'Address not found or invalid'
      });
    }
  } catch (error) {
    console.error('Address validation error:', error);
    res.status(500).json({ message: 'Error validating address' });
  }
});

// Get address from coordinates (reverse geocoding)
router.post('/reverse-geocode', protect, async (req, res) => {
  try {
    const { latitude, longitude } = req.body;
    
    if (!latitude || !longitude) {
      return res.status(400).json({ message: 'Latitude and longitude are required' });
    }

    const apiKey = process.env.GOOGLE_MAPS_API_KEY;
    if (!apiKey) {
      return res.status(500).json({ message: 'Google Maps API key not configured' });
    }

    const response = await fetch(
      `https://maps.googleapis.com/maps/api/geocode/json?latlng=${latitude},${longitude}&key=${apiKey}`
    );
    
    const data = await response.json();
    
    if (data.status === 'OK' && data.results.length > 0) {
      const result = data.results[0];
      const addressComponents = result.address_components;
      
      const streetNumber = addressComponents.find(component => 
        component.types.includes('street_number')
      )?.long_name || '';
      
      const route = addressComponents.find(component => 
        component.types.includes('route')
      )?.long_name || '';
      
      const locality = addressComponents.find(component => 
        component.types.includes('locality')
      )?.long_name || '';
      
      const postalCode = addressComponents.find(component => 
        component.types.includes('postal_code')
      )?.long_name || '';
      
      const country = addressComponents.find(component => 
        component.types.includes('country')
      )?.long_name || '';
      
      res.json({
        formattedAddress: result.formatted_address,
        addressComponents: {
          street: `${streetNumber} ${route}`.trim(),
          city: locality,
          pincode: postalCode,
          country: country
        }
      });
    } else {
      res.status(404).json({ message: 'Address not found for these coordinates' });
    }
  } catch (error) {
    console.error('Reverse geocoding error:', error);
    res.status(500).json({ message: 'Error getting address from coordinates' });
  }
});

module.exports = router; 

File: backend-python/routes/address.py
from fastapi import APIRouter, HTTPException, status, Depends, Query
from motor.motor_asyncio import AsyncIOMotorClient
import logging
import httpx
import os
from typing import Optional

from utils.database import DatabaseManager, get_database
from utils.auth import get_current_active_user
from models.user import UserInDB

logger = logging.getLogger(__name__)
router = APIRouter()

@router.get("/search")
async def search_addresses(
    query: str = Query(..., description="Search query for addresses"),
    db: DatabaseManager = Depends(get_database)
):
    """Search addresses using Google Maps Places Autocomplete API"""
    try:
        if not query:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Query parameter is required"
            )

        api_key = os.getenv("GOOGLE_MAPS_API_KEY")
        if not api_key:
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Google Maps API key not configured"
            )

        # Use Places Autocomplete API
        async with httpx.AsyncClient() as client:
            response = await client.get(
                "https://maps.googleapis.com/maps/api/place/autocomplete/json",
                params={
                    "input": query,
                    "key": api_key,
                    "components": "country:in"
                }
            )
            
            data = response.json()
            
            if data.get("status") == "OK":
                return {"predictions": data.get("predictions", [])}
            else:
                return {"predictions": []}
                
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Address search error: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Error searching addresses"
        )

@router.post("/geocode")
async def geocode_address(
    address: str,
    db: DatabaseManager = Depends(get_database)
):
    """Geocode address to get coordinates"""
    try:
        if not address:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Address is required"
            )

        api_key = os.getenv("GOOGLE_MAPS_API_KEY")
        if not api_key:
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Google Maps API key not configured"
            )

        # Geocode the address
        async with httpx.AsyncClient() as client:
            response = await client.get(
                "https://maps.googleapis.com/maps/api/geocode/json",
                params={
                    "address": address,
                    "key": api_key
                }
            )
            
            data = response.json()
            
            if data.get("status") == "OK" and data.get("results"):
                result = data["results"][0]
                location = result["geometry"]["location"]
                
                return {
                    "latitude": location["lat"],
                    "longitude": location["lng"],
                    "formattedAddress": result["formatted_address"]
                }
            else:
                raise HTTPException(
                    status_code=status.HTTP_404_NOT_FOUND,
                    detail="Address not found"
                )
                
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Geocoding error: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Error geocoding address"
        )

@router.post("/validate")
async def validate_address(
    address: str,
    current_user: UserInDB = Depends(get_current_active_user),
    db: DatabaseManager = Depends(get_database)
):
    """Validate address using Google Maps Geocoding API"""
    try:
        if not address:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Address is required"
            )

        api_key = os.getenv("GOOGLE_MAPS_API_KEY")
        if not api_key:
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Google Maps API key not configured"
            )

        # Geocode the address
        async with httpx.AsyncClient() as client:
            response = await client.get(
                "https://maps.googleapis.com/maps/api/geocode/json",
                params={
                    "address": address,
                    "key": api_key
                }
            )
            
            data = response.json()
            
            if data.get("status") == "OK" and data.get("results"):
                result = data["results"][0]
                location = result["geometry"]["location"]
                
                # Extract address components
                address_components = result["address_components"]
                
                def find_component(types_list):
                    for component in address_components:
                        if any(t in component["types"] for t in types_list):
                            return component["long_name"]
                    return ""
                
                street_number = find_component(["street_number"])
                route = find_component(["route"])
                locality = find_component(["locality"])
                postal_code = find_component(["postal_code"])
                country = find_component(["country"])
                
                return {
                    "valid": True,
                    "coordinates": {
                        "latitude": location["lat"],
                        "longitude": location["lng"]
                    },
                    "formattedAddress": result["formatted_address"],
                    "addressComponents": {
                        "street": f"{street_number} {route}".strip(),
                        "city": locality,
                        "pincode": postal_code,
                        "country": country
                    }
                }
            else:
                return {
                    "valid": False,
                    "message": "Address not found or invalid"
                }
                
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Address validation error: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Error validating address"
        )

@router.post("/reverse-geocode")
async def reverse_geocode(
    latitude: float,
    longitude: float,
    current_user: UserInDB = Depends(get_current_active_user),
    db: DatabaseManager = Depends(get_database)
):
    """Get address from coordinates (reverse geocoding)"""
    try:
        if not latitude or not longitude:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Latitude and longitude are required"
            )

        api_key = os.getenv("GOOGLE_MAPS_API_KEY")
        if not api_key:
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Google Maps API key not configured"
            )

        async with httpx.AsyncClient() as client:
            response = await client.get(
                "https://maps.googleapis.com/maps/api/geocode/json",
                params={
                    "latlng": f"{latitude},{longitude}",
                    "key": api_key
                }
            )
            
            data = response.json()
            
            if data.get("status") == "OK" and data.get("results"):
                result = data["results"][0]
                address_components = result["address_components"]
                
                def find_component(types_list):
                    for component in address_components:
                        if any(t in component["types"] for t in types_list):
                            return component["long_name"]
                    return ""
                
                street_number = find_component(["street_number"])
                route = find_component(["route"])
                locality = find_component(["locality"])
                postal_code = find_component(["postal_code"])
                country = find_component(["country"])
                
                return {
                    "formattedAddress": result["formatted_address"],
                    "addressComponents": {
                        "street": f"{street_number} {route}".strip(),
                        "city": locality,
                        "pincode": postal_code,
                        "country": country
                    }
                }
            else:
                raise HTTPException(
                    status_code=status.HTTP_404_NOT_FOUND,
                    detail="Address not found for these coordinates"
                )
                
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Reverse geocoding error: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Error getting address from coordinates"
        ) 